### PR TITLE
[OPENJDK-2731] Behave test for jlink integration code

### PR DESCRIPTION
**Pull request by jhuttana**
Could you please review ?

### CODE REVIEW

File: modules/jlink/tests/features/jlink.feature [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+Feature: Openshift OpenJDK S2I tests (jlink specific)
+Scenario: Ensure jlinked builder is used to build the containerized application image
+      Given s2i build https://github.com/jboss-container-images/openjdk-test-applications from quarkus-quickstarts/getting-started-3.0.1.Final-nos2i
+       | variable            | value        |
+       | S2I_ENABLE_JLINK    | true         |
+       | QUARKUS_PACKAGE_TYPE| uber-jar     |
+      Then run ls /tmp/jre in container and check its output for bin
+
+Scenario: Ensure S2I_ENABLE_JLINK=false is the default
+      Given container is started with env
```
Comment (jmtd) [inline modules/jlink/tests/features/jlink.feature]: this scenario won't need an environment table

Comment (jmtd) [inline modules/jlink/tests/features/jlink.feature]: for this test to work the container needs to run the command `env`, see [this step](https://github.com/cekit/behave-test-steps/blob/v1/steps/container_steps.py#L106)

**Comment by jhuttana** [comment]
I have pushed my recent changes to address the review comments but the latest commit is not showing up in this PR!!
Whereas I can see the commit on my fork-branch : https://github.com/jhuttana/openjdk/commits/jlinkcode_integration/

**Comment by jhuttana** [comment]
> I have pushed my recent changes to address the review comments but the latest commit is not showing up in this PR!! Whereas I can see the commit on my fork-branch : https://github.com/jhuttana/openjdk/commits/jlinkcode_integration/

Meanwhile I was trying some hacks like changing the base etc and then the latest commit appeared :D Not sure what worked but its fine now :)

**Comment by jmtd** [comment]
Hi Jaya, I've been thinking about `S2I_ENABLE_JLINK` and I think we should alter it so that it is NOT defined by default (to true or false) and so meets Cekit's definition of a "configuration environment variable". I've filed [OPENJDK-2806](https://issues.redhat.com/browse/OPENJDK-2806) (S2I_ENABLE_JLINK should be a configuration env var only) and PR #453, but it  means changes to this PR too.

I think the second test is OK, but should be renamed to describe that it is not testing for `S2I_ENABLE_JLINK=false`, but that it's _not_ `S2I_ENABLE_JLINK=true`;

The first test is OK, but for completion can we also have another which performs the inverse? I.e., when `S2I_ENABLE_JLINK` is _not_ set by the user, test that `/tmp/jre` does not exist.

**Comment by jmtd** [comment]
Please also add a test for the JDK8 image that proves that `S2I_ENABLE_JLINK=true` does not work for that image; My changes in #449 should make this a fatal error so we might need to experiment with the test to figure out how to catch that behaviour.

**Comment by jhuttana** [comment]
> Hi Jaya, I've been thinking about `S2I_ENABLE_JLINK` and I think we should alter it so that it is NOT defined by default (to true or false) and so meets Cekit's definition of a "configuration environment variable". I've filed [OPENJDK-2806](https://issues.redhat.com/browse/OPENJDK-2806) (S2I_ENABLE_JLINK should be a configuration env var only) and PR #453, but it means changes to this PR too.
> 
> I think the second test is OK, but should be renamed to describe that it is not testing for `S2I_ENABLE_JLINK=false`, but that it's _not_ `S2I_ENABLE_JLINK=true`;

+1  makes more sense 

> 
> The first test is OK, but for completion can we also have another which performs the inverse? I.e., when `S2I_ENABLE_JLINK` is _not_ set by the user, test that `/tmp/jre` does not exist.

If my understanding is right then, logically  `S2I_ENABLE_JLINK` is not set and `S2I_ENABLE_JLINK=false` are both same. 
In both  cases `/tmp/jre`  is not created as it should not take the jlink path. So if at all we want to test `/tmp/jre` doesn't exist then we can club it with second Scenario. WDYT?

File: modules/jlink/tests/features/jlink.feature [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      Given s2i build https://github.com/jboss-container-images/openjdk-test-applications from quarkus-quickstarts/getting-started-3.0.1.Final-nos2i
+       | variable            | value        |
+       | S2I_ENABLE_JLINK    | true         |
+       | QUARKUS_PACKAGE_TYPE| uber-jar     |
+      Then run ls /tmp/jre in container and check its output for bin
+
+Scenario: Ensure S2I_ENABLE_JLINK is not set to true
+      Given s2i build https://github.com/jboss-container-images/openjdk-test-applications from quarkus-quickstarts/getting-started-3.0.1.Final-nos2i
+      Then container log should not contain S2I_ENABLE_JLINK=true
+      And run ls /tmp/jre in container and check its output for ls:
```
Comment (jmtd) [inline modules/jlink/tests/features/jlink.feature]: I guess `ls:` here is a string that ls prints when it doesn't find something, but it's not clear. Could we look for `No such file or directory` instead?

**Comment by jmtd** [comment]
>  In both cases /tmp/jre is not created as it should not take the jlink path. So if at all we want to test /tmp/jre doesn't exist then we can club it with second Scenario. WDYT?

Yes that makes sense.

File: modules/jlink/tests/features/jlink.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      Given s2i build https://github.com/jboss-container-images/openjdk-test-applications from quarkus-quickstarts/getting-started-3.0.1.Final-nos2i
+       | variable            | value        |
+       | S2I_ENABLE_JLINK    | true         |
+       | QUARKUS_PACKAGE_TYPE| uber-jar     |
+      Then run ls /tmp/jre in container and check its output for bin
+
+Scenario: Ensure S2I_ENABLE_JLINK is not set to true
+      Given s2i build https://github.com/jboss-container-images/openjdk-test-applications from quarkus-quickstarts/getting-started-3.0.1.Final-nos2i
+      Then container log should not contain S2I_ENABLE_JLINK=true
+      And run ls /tmp/jre in container and check its output for No such file or directory
```
Comment (jmtd) [inline modules/jlink/tests/features/jlink.feature]: I'm not able to get this test to pass when it should locally. I'm not sure why: stderr output not in container log? case sensitivity? locale sensitivity for the format of the ls error output? I see this:
`      Exception: ("Phrase 'No such file or directory' was not found in the output of running the 'ls /tmp/jre' command", b"ls: cannot access '/tmp/jre': No such file or directory\n", ExecException('Command ls /tmp/jre failed to execute, return code: 2'))
`

Comment (jmtd) [reply modules/jlink/tests/features/jlink.feature]: I've come up with an alternative solution, I'll push it

**Comment by jhuttana** [comment]
Thank you :)

**Pull request merged**